base/applications/rapps/unattended.cpp

```diff
@@ -312,6 +313,48 @@ HandleInfoCommand(CAppDB *db, LPWSTR szCommand, int argcLeft, LPWSTR *argvLeft)
     return TRUE;
 }
 
+static unsigned int CALLBACK
+RunGuiThread(void *nCmdShow)
+{
+    WCHAR fakecmdline[] = L"://";
+    return ParseCmdAndExecute(fakecmdline, PCAEF_WINDOWONLY, (UINT)(SIZE_T)nCmdShow);
+}
+
+static BOOL
+HandleProtocolCommand(CAppDB &db, LPWSTR Url, int nCmdShow)
```

Comment:
It should be possible to have `Url` as a `LPCWSTR`.

Reply:
No, I'm writing to it

Reply:
You mean, the UrlUnescapeInPlace() call above _only_ in the `L"install/"` case below?
If so, why not just make a local string capture as you did at the beginning of `CMainWindow::HandleProtocolMessage` ?

Reply:
Why would I make a copy if I don't have to?

> If so, why not just make a local string capture as you did at the beginning of CMainWindow::HandleProtocolMessage ?

The only reason I made a copy there is that I don't know if I'm allowed to write to the WM_COPYDATA memory. It does work but I don't know if it would somehow be possible for it to not be R/W in other cases/versions.

Reply:
This smells like a breaking of abstraction, that a function that takes in parameter something, and does an action that should not modify the parameter, actually does modify the parameter as a unwanted side-effect, from the point of view of the person who wants to use that function.
(It's like CreateProcessW)


>The only reason I made a copy there is that I don't know if I'm allowed to write to the WM_COPYDATA memory. It does work but I don't know if it would somehow be possible for it to not be R/W in other cases/versions.

https://learn.microsoft.com/en-us/windows/win32/dataxchg/wm-copydata#remarks

> While this message is being sent, the referenced data must not be changed by another thread of the sending process.
The receiving application should consider the data read-only. The lParam parameter is valid only during the processing of the message. The receiving application should not free the memory referenced by lParam. If the receiving application must access the data after [SendMessage](https://learn.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-sendmessage) returns, it must copy the data into a local buffer.

If you were writing into the WM_COPYDATA memory you would just rely on some undefined behaviour, that could cause a crash or whatnot.

Reply:
In general, yes, but these two functions are pretty dependent on each other.

It is not unheard of to change the command line, pretty common on *NIX.